codePost_lib.py

```diff
@@ -0,0 +1,695 @@
+#
+##########################################################################
+
+import sys as _sys
+import requests as _requests
+
+try:
+    # Python 2
+    from urllib import quote as _urlquote
+except ImportError:
+    # Python 3
+    from urllib.parse import quote as _urlquote
+
+BASE_URL = 'https://api.codepost.io'
+
+##########################################################################
+
+
+class _color:
+    RED = '\033[91m'
+    BOLD = '\033[1m'
+    UNDERLINE = '\033[4m'
+    END = '\033[0m'
+
+
+_TERM_INFO = "{END}[{BOLD}INFO{END}]{END}".format(**_color.__dict__)
+_TERM_ERROR = "{END}[{BOLD}{RED}ERROR{END}]{END}".format(**_color.__dict__)
+
+
+def _print_info(msg):
+    print("{tag} {msg}".format(tag=_TERM_INFO, msg=msg), file=_sys.stderr)
+
+
+DEFAULT_MODE = {
+    "updateIfExists": True,
+    "updateIfClaimed": True,
+    "resolveStudents": True,
+
+    "addFiles": True,
+    "updateExistingFiles": True,
+
+    "removeComments": True,
+    "doUnclaim": True,
+    "deleteAffectedSubmissions": True
+}
+
+###########################################################################################
+
+
+class UploadError(RuntimeError):
+    pass
+
+
+def upload_submission(api_key, assignment, students, files, mode=DEFAULT_MODE):
+
+    assignment_id = assignment.get("id", 0)
+
+    # Retrieve all existing submissions associated with the students
+
+    existing_submissions = {}
+
+    for student in students:
+        submissions = get_assignment_submissions(
+            api_key=api_key,
+            assignment_id=assignment_id,
+            username=student
+        )
+
+        for submission in submissions:
+            existing_submissions[submission["id"]] = submission
+
+    # Check to see if there is a collision
+
+    if len(existing_submissions) == 0:
+
+        # CASE 1: No existing submission => create a new submission
+
+        return post_submission(
+            api_key=api_key,
+            assignment_id=assignment_id,
+            students=students,
+            files=files
+        )
+
+    # There is at least one (maybe more) existing submissions
+
+    # First check the modes to determine whether to proceed.
+    if not mode["updateIfExists"]:
+        raise UploadError(
+            """
+            At least one submission already exists, and 'updateIfExists' is false,
+            so interrupting upload.
+            """)
+
+    # Check whether any of the existing submissions are claimed.
+    if not mode["updateIfClaimed"] and _submission_list_is_unclaimed(existing_submissions):
+        raise UploadError(
+            """
+            At least one submission has already been claimed by a grader, and
+            'updateIfClaimed' is false, so interrupting upload.
+            """)
+
+    # Check whether students will need an update.
+    if not mode["resolveStudents"]:
+        if set(students) != set(existing_submissions[0]["students"]) or len(existing_submissions) > 1:
```

Comment:
This boolean is very elegant 👏

Reply:
You can slightly improve performance by switching the order of the clauses. If `len(existing_submissions) > 1`, then we don't need to construct a set out of `existing_submissions[0]["students"]`.

Reply:
Thanks! And thanks for the improvement.